### DISCUSSION TITLE
next/image does not seem to follow redirects

### DISCUSSION THREAD

**Discussion by dislick** [vercel/next.js#36808] [Help]
I've been experimenting with next/image and found that it does not follow HTTP redirects. Here's what I'd like to accomplish:

An API route `/api/images/[imageId]` responds with a redirect:

```
GET /api/images/123
```

```
HTTP/1.1 302 Found
Connection: keep-alive
Date: Tue, 10 May 2022 14:15:06 GMT
Keep-Alive: timeout=5
Location: https://redacted.s3.eu-central-1.amazonaws.com/redacted-signed-url-goes-here
Transfer-Encoding: chunked
```

The `/api/images/[imagedId]` route creates a [signed URL to read an AWS S3 object](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html). It then redirects to said URL with the `Location` header. This works fine in a regular `<img>` tag. As soon as I try to use it with `next/image` though, it fails. This is the HTTP response I'm getting:

```
GET /_next/image\?url\=%2Fapi%2Fimages%2F123\&w\=640\&q\=75
```

```
HTTP/1.1 302 Found
Cache-Control: public, max-age=0, must-revalidate
Connection: keep-alive
Content-Security-Policy: script-src 'none'; sandbox;
Date: Tue, 10 May 2022 14:17:24 GMT
ETag: 47DEQpj8HBSa+-TImW+5JCeuQeRkm5NMpJWZG3hSuFU=
Keep-Alive: timeout=5
Transfer-Encoding: chunked
Vary: Accept
```

The body is empty. `<Image>` does not render anything and there are no logs in the console. Is this intended behaviour? Why isn't the redirect being followed? Am I doing something wrong?

**Why do I even want this?**

My intention is to have a constant URL that can be used as a cache key by `next/image` that then returns a temporary signed URL to S3. This way I don't need to make the S3 objects public. If I would just pass the signed URL to `next/image`, they would be different for each request and the caching mechanism would be broken.

**Comment by spoldman** [comment]
I'm having the same problem as redirects using a API route does not seam work when using Next image component. 
Looking at the console I'm also getting "API resolved without sending a response for /api/xxx, this may result in stalled requests" something I never get when using regular img tag. So there is something happening behind the schenes when using this that is hard to figure out or debug.

**Reply by noce2** [reply]
Having the same issue here too!

**Comment by pencildrummer** [comment]
Same issue here.

I found a temporary solution, just set `unoptimized` on the `<Image>` component.

The issue seems to be related to the `image-optimizer.js` module.
The image optimizer ignores the status code when it's a redirect and thus check the Content-Type even for redirect requests.

If interested should be around line 380 in image-optimizer.js:
```
try {
            const { resBuffers , req: mockReq , res: mockRes , streamPromise: isStreamFinished ,  } = (0, _mockRequest).mockRequest(href, _req.headers, _req.method || "GET", _req.connection);
            await handleRequest(mockReq, mockRes, _url.default.parse(href, true));
            await isStreamFinished;
            if (!mockRes.statusCode) {
                console.error("image response failed for", href, mockRes.statusCode);
                throw new ImageError(mockRes.statusCode, '"url" parameter is valid but internal response is invalid');
            }
            upstreamBuffer = Buffer.concat(resBuffers);
            upstreamType = detectContentType(upstreamBuffer) || mockRes.getHeader("Content-Type");
            maxAge = getMaxAge(mockRes.getHeader("Cache-Control"));
        } catch (err) {
            console.error("upstream image response failed for", href, err);
            throw new ImageError(500, '"url" parameter is valid but upstream response is invalid');
        }
```

**Comment by quintus-luis** [comment]
Same issue here! Would be great if this would work! Because using  `unoptimized` kinda makes the whole next/image obsolete

**Comment by dylankbuckley** [comment]
Also an issue - setup a redirect in nextconfig - works fine with `unoptimized ` but otherwise not

**Comment by jankocian** [comment]
Any update on this one? 

I'm facing the same issue and it is pretty cumbersome. The API route redirect to a presigned url seems like an elegant solution for media on S3.

Would you recommend a good workaround?

**Comment by jneterer** [comment]
Hello. Would like to comment this is an issue for our repositories as well. We have a redirect set up for images and this is preventing us from using the NextImage component.

**Comment by sgoodrow** [comment]
Bump, also looking for a solution like this. Thank!

**Comment by ayrtoneverton** [comment]
I'm going to share an alternative solution that I found and I believe it will help a lot of people...

As already mentioned, the problem is not actually in the `Image` component, but rather in the NextJS image optimization endpoint that this component uses, which is `/_next/image?url=ANY-URL&w=1080&q=75 `.

From what I noticed in the tests I did, it can't correctly understand redirects when using just URI in the `url` parameter, so something like this doesn't work:
```jsx
❌ <Image src="/go-img/img-id" width={1080} height={500} alt="" />
```
```url
❌ http://localhost:3000/_next/image?url=/go-img/img-id&w=1080&q=75
```

But I noticed that when using the full URL in the parameter, it works:
```jsx
✅ <Image src="http://localhost:3000/go-img/img-id" width={1080} height={500} alt="" />
```
```url
✅ http://localhost:3000/_next/image?url=http://localhost:3000/go-img/img-id&w=1080&q=75
```

⚠️ **ALERT:** Note that this will make it necessary to add your own project domain to `images.remotePatterns`([Doc](https://nextjs.org/docs/pages/api-reference/components/image#remotepatterns)) of `next.config.mjs`. And be very careful not to leave `localhost` or something like that released in your production environments.

Extra tip: If the images were public, it is not necessary to create an endpoint to do this type of redirection, it can be done dynamically through `next.config.mjs` ([Doc](https://nextjs.org/docs/pages/api-reference/next-config-js/redirects#path-matching)), for example:
```js
const nextConfig = {
  images: {
    remotePatterns: [
      { hostname: `your-domain.com` },
      process.env.NODE_ENV !== 'production' && { hostname: 'localhost' },
    ].filter(Boolean),
  },
  redirects: () => [{
    source: '/go-file/:context/:fileId',
    destination: `https://${process.env.AWS_BUCKET_NAME}.s3.${process.env.AWS_REGION}.amazonaws.com/:context/:fileId`,
    permanent: true,
  }],
};
```

I hope this helps! :)

**Reply by backlands** [reply]
For those that are performing this functionality for the purpose of presigned keys from S3 you do have another option. You can instead create a middleware that rewrites requests for S3 assets to a presigned key generated on request. Unfortunately this doesn't work if you are seeking to create presigned keys for CloudFront assets as that signing method requires `crypto` which is unavailable in the edge runtime used for middlewares.

For reference if you are doing this for S3 assets, here is a reproducible setup:

In your image components prepend all S3 assets with a unique path, such as "/storage/":

```tsx
<Image src={`/storage/${url}`} />
```

Create a middleware handler to handle the `/storage` path (`middleware.ts`):

```ts
import { NextRequest, NextResponse } from 'next/server';

import { loadAssetFromStorage } from './utilities/storage';

export async function middleware(request: NextRequest) {
  const path = request.nextUrl.pathname;

  // If the path is loaded from "storage" we pass it to our asset loader to add presigned keys
  if (path.startsWith('/storage/')) return NextResponse.rewrite(await loadAssetFromStorage(path));

  ...add other middlewares you might have

  return NextResponse.next();
}

export const config = {
  matcher: [
    // Skip all internal paths (_next) to prevent breaking chunking on pages
    '/((?!_next|api|static|public|favicon.ico).*)',
  ],
};
```

Then you create a handler for `loadAssetFromStorage` which generates the presigned key for the requested storage path:

```ts
import {
  S3Client,
  GetObjectCommand,
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';

const getPresignedDownloadUrl = async (
  client: S3Client,
  key: string,
) => {
  const command = new GetObjectCommand({
    Bucket: process.env.AWS_BUCKET,
    Key: key,
  });

  return getSignedUrl(client, command, { expiresIn: 3600 });
};

const loadAssetFromStorage = async (path: string): Promise<string> => {
  const prefix = '/storage/';

  const key = path.slice(prefix.length);

  const client = getStorageClient();

  return getPresignedDownloadUrl(client, key);
};
```

Doing this will help you avoid the above process for bypassing the redirect limitation of the image optimization tooling as it appears that a rewrite from middleware is respected and followed, unlike a redirect.

**Comment by boylec** [comment]
I'm wondering, when rewriting image URLs to an external domain (such as a CDN or CMS endpoint), does the remotePatterns need to be setup to allow image optimization for those images requested from the external domain due to the rewrite?

I guess simpler way to put it: does image optimization logic for whether to optimize or not key off of the rewritten URL or the pre-rewritten URL?
